### PR TITLE
[Core] Add a charspan type to CC dialect

### PR DESCRIPTION
**Pull request by schweitzpgi**
Added support for std::string to the frontend as well.

Supports #1382

### CODE REVIEW

File: include/cudaq/Optimizer/Dialect/CC/CCTypes.td [hunk 1]
```diff
@@ -226,4 +227,31 @@ def cc_StdVectorType : CCType<"Stdvec", "stdvec"> {
...
 }
 
+def cc_CharSpanType
+    : CCType<"Charspan", "charspan", [], "cudaq::cc::SpanLikeType"> {
+  let summary = "Proxy for a span over a std::string in cc";
+  let description = [{
+    This captures a C++ std::string(of char) type as an abstract CC type. The
+    intention is to allow us to reason about std::string at the CC level before
+    decomposing these into more refined data structures
+                .
```
Comment (amccaskey) [inline include/cudaq/Optimizer/Dialect/CC/CCTypes.td:237]: should this period be all the way out here?

File: include/cudaq/Optimizer/Dialect/CC/CCTypes.td [hunk 2]
```diff
@@ -226,4 +227,31 @@ def cc_StdVectorType : CCType<"Stdvec", "stdvec"> {
...
+
+  let extraClassDeclaration = [{
+    mlir::IntegerType getElementType() const {
+      return mlir::IntegerType::get(getContext(), 8);
+    }
+  }];
+}
+
+def IsSpanLikePred : CPred<"$_self.isa<cudaq::cc::SpanlikeType>()>">;
+def fir_SpanLikeType : Type<IsSpanLikePred, "dynamic length type">;
```
Comment (amccaskey) [inline include/cudaq/Optimizer/Dialect/CC/CCTypes.td:255]: old `fir` prefix here

File: lib/Optimizer/CodeGen/LowerToQIR.cpp [hunk 3]
```diff
@@ -1652,6 +1652,10 @@ void cudaq::opt::initializeTypeConversions(LLVMTypeConverter &typeConverter) {
     auto eleTy = typeConverter.convertType(type.getElementType());
     return factory::stdVectorImplType(eleTy);
   });
+  typeConverter.addConversion([&typeConverter](cc::CharspanType type) {
+    auto eleTy = typeConverter.convertType(type.getElementType());
+    return factory::stdStringImplType(eleTy);
```
Comment (amccaskey) [inline lib/Optimizer/CodeGen/LowerToQIR.cpp:1657]: Why use `stdStringImplType` or why does `stdStringImplType` return a `{elePtrTy, i64Ty, i64Ty, i64Ty}`? Shouldn't a span type just be `{elePtrTy, i64Ty}`?

File: include/cudaq/Optimizer/Builder/Factory.h [hunk 4]
```diff
@@ -107,6 +107,14 @@ inline mlir::LLVM::LLVMStructType stdVectorImplType(mlir::Type eleTy) {
   return mlir::LLVM::LLVMStructType::getLiteral(ctx, eleTys);
 }
 
+inline mlir::LLVM::LLVMStructType stdStringImplType(mlir::Type eleTy) {
+  auto *ctx = eleTy.getContext();
+  auto elePtrTy = cudaq::opt::factory::getPointerType(eleTy);
+  auto i64Ty = mlir::IntegerType::get(ctx, 64);
+  llvm::SmallVector<mlir::Type> eleTys = {elePtrTy, i64Ty, i64Ty, i64Ty};
```
Comment (amccaskey) [inline include/cudaq/Optimizer/Builder/Factory.h:114]: I thought we were going for span types to be `{T*, i64}`?

Comment (schweitzpgi) [reply include/cudaq/Optimizer/Builder/Factory.h:114]: Yes, we are.

File: include/cudaq/Optimizer/Dialect/CC/CCTypes.td (hunk 2 above)
Comment (schweitzpgi) [reply include/cudaq/Optimizer/Dialect/CC/CCTypes.td:255]: oops. Old habits die hard.

File: lib/Optimizer/CodeGen/LowerToQIR.cpp (hunk 3 above)
Comment (schweitzpgi) [reply lib/Optimizer/CodeGen/LowerToQIR.cpp:1657]: My bad. We should be using the span form here. Doh!

File: lib/Optimizer/Builder/Factory.cpp [hunk 5]
```diff
@@ -225,19 +225,27 @@ bool factory::hasHiddenSRet(FunctionType funcTy) {
...
-                                 .isa<cc::StdvecType, cc::StructType,
+                                 .isa<cc::SpanLikeType, cc::StructType,
                                       cc::ArrayType, cc::CallableType>());
 }
 
+cc::StructType factory::stlStringType(MLIRContext *ctx) {
+  auto i8Ty = IntegerType::get(ctx, 8);
+  auto ptrI8Ty = cc::PointerType::get(i8Ty);
+  auto i64Ty = IntegerType::get(ctx, 64);
+  auto padTy = cc::ArrayType::get(ctx, i8Ty, 16);
```
Comment (1tnguyen) [inline lib/Optimizer/Builder/Factory.cpp:236]: Would it make sense to add a `static_assert` here to validate our layout assumption?
e.g.,
```
static_assert(sizeof(std::string) == sizeof(char*) + sizeof(std::size_t) + 16);
```

**Review by 1tnguyen — APPROVED** [review]
LGTM 👍

**Pull request merged**